### PR TITLE
[PATCH v7] random data improvements and tests

### PR DESCRIPTION
**Pull request by JereLeppanen**
```
linux-gen: random: improve ODP_RANDOM_BASIC data and random test data
    
    Improve the quality of data generated by and the speed of
    odp_random_test_data() and _odp_random_std_data() (used when OpenSSL
    random is not enabled).
    
    Signed-off-by: Jere Leppänen <jere.leppanen@nokia.com>

validation: random: add test for buffer overflow and alignment problems
    
    Add a new test, which checks that odp_random_data() and
    odp_random_test_data() do not modify bytes outside the given buffer.
    
    Signed-off-by: Jere Leppänen <jere.leppanen@nokia.com>
    Reviewed-by: Petri Savolainen <petri.savolainen@nokia.com>

validation: random: test randomness
    
    Test that random data looks random by adding four new tests:
    
    random_test_frequency checks that random data is uniformly
    distributed.
    
    random_test_independence checks that bytes are independent.
    
    random_test_runs checks that the number of runs (uninterrupted
    sequence of identical bits) is what is expected of random data.
    
    random_test_matrix_rank checks that the ranks of 32x32 binary matrices
    correspond to expected probabilities.
    
    Signed-off-by: Jere Leppänen <jere.leppanen@nokia.com>
```
v2:
- Tuomas' comments.

v3:
- Rebase.
- Remove API change (separate it into its own PR).

v4:
- Add review tags.

v6:
- Rebase.
- Remove commit "validation: chksum: use odp_random_test_data()".
- Petri's comments.
- Add review tags.

### CODE REVIEW

File: platform/linux-generic/odp_random_std.c [hunk 1]
```diff
@@ -4,52 +4,88 @@
...
+			*(uint16_t *)(uintptr_t)buf = r & 0xffff;
+			r >>= 16;
+			buf += 2;
+		}
+
+		if (len & 1)
+			*(uint8_t *)(uintptr_t)buf = r & 0xff;
+	}
+
+	return ret;
```
Comment (TuomasTaipale) [inline platform/linux-generic/odp_random_std.c:80]: Is `memcpy()` too slow here (would handle alignment issues and there would be less code)?

File: test/validation/api/random/random.c [hunk 2]
```diff
@@ -80,10 +80,370 @@ static void random_test_repeat(void)
...
+ * Perform randomness tests on this and higher types of random data. To test
+ * also odp_random_test_data(), set MIN_KIND to -1.
+ */
+#define MIN_KIND ODP_RANDOM_CRYPTO
+
+/*
+ * Alpha for P-value tests. This does not affect the tests that use a
+ * precomputed critical value.
+ */
+const double alpha = 0.00000001;
```
Comment (TuomasTaipale) [inline test/validation/api/random/random.c:105]: This could be `static`.

File: test/validation/api/random/random.c [hunk 3]
```diff
@@ -80,10 +80,370 @@ static void random_test_repeat(void)
...
+			r |= buf[bit / 8];
+			bit += 8;
+			i += 8;
+			continue;
+		}
+		/* Single bit. */
+		r <<= 1;
+		r |= (buf[bit / 8] >> (7 - (bit & 7))) & 1;
+		bit++;
+		i++;
```
Comment (TuomasTaipale) [inline test/validation/api/random/random.c:160]: Pre-increment is typically slightly faster as there's no need for the implicit temporary variable.

File: test/validation/api/random/random.c [hunk 4]
```diff
@@ -80,10 +80,370 @@ static void random_test_repeat(void)
...
+				chisq += dif * dif / expected;
+			}
+
+			printf("%s ; bits %d ; chisq %g ; df %u ; crit %g ; ",
+			       kname(kind), bits, chisq, cells - 1, crit);
+
+			if (chisq < crit)
+				printf("pass\n");
+			else
+				printf("FAIL\n");
```
Comment (TuomasTaipale) [inline test/validation/api/random/random.c:223]: A bit less code:
```c
printf("%s\n", chisq < crit ? "pass" : "FAIL");
```

File: platform/linux-generic/odp_random_std.c (hunk 1 above)
Comment (JereLeppanen) [reply platform/linux-generic/odp_random_std.c:80]: Unfortunately memcpy() is quite a lot slower. In the worst case, around 20% lower throughput in the performance test (in my environment).

File: test/validation/api/random/random.c (hunk 3 above)
Comment (JereLeppanen) [reply test/validation/api/random/random.c:160]: When it's a basic type, the unused temporary is usually optimized away completely, and the resulting code from pre-increment and post-increment is exactly the same. At least gcc does this even with -O0. The convention seems to be to use post-increment unless there's a specific reason to use pre-increment. I can get behind that, because pre-increment looks ugly.

File: test/validation/api/random/random.c (hunk 4 above)
Comment (JereLeppanen) [reply test/validation/api/random/random.c]: I put this in a function to boot, so that the strings appear in just one place.

File: test/validation/api/random/random.c (hunk 3 above)
Comment (TuomasTaipale) [reply test/validation/api/random/random.c:160]: OK. Well, it's trusting that compiler makes it fast vs. making it fast by default. In any case, doesn't probably matter here.

**Review by TuomasTaipale — APPROVED** [review]

File: test/validation/api/random/random.c [hunk 5]
```diff
@@ -80,10 +80,357 @@ static void random_test_repeat(void)
...
+	const double critical[] = {
+		32.8413, 40.1300, 50.8129, 68.0293,
+		97.0285, 147.463, 237.614, 402.685,
+		711.187, 1297.50, 2426.64, 4623.37,
+		8929.74, 17419.3, 34224.0, 67587.1,
+	};
+
+	printf("\n\n");
+
+	for (int kind = MIN_KIND; kind <= (int)odp_random_max_kind(); kind++) {
```
Comment (psavol) [inline test/validation/api/random/random.c:205]: I'd prefer random kinds to be tested in separate test cases. Implementation may have different implementation for different kinds, and some of those could fail randomness test more often than others. As separate cases, it would be easier to track which kinds were tested (which kind failed a case) when those are separate cases.

File: platform/linux-generic/odp_random_std.c [hunk 6]
```diff
@@ -4,52 +4,88 @@
...
+			*(uint8_t *)(uintptr_t)buf = r & 0xff;
+			r >>= 8;
+			buf += 1;
+			len -= 1;
+		}
+
+		if ((uintptr_t)buf & 2) {
+			*(uint16_t *)(uintptr_t)buf = r & 0xffff;
+			buf += 2;
+			len -= 2;
```
Comment (psavol) [inline platform/linux-generic/odp_random_std.c:58]: Len underflows here if application requests a small number of bytes (e.g. len = 1).

A test case could be added to test various small lengths (and unaligned buf pointers).

File: test/validation/api/random/random.c [hunk 7]
```diff
@@ -80,10 +80,84 @@ static void random_test_repeat(void)
...
+				(int32_t)len);
+	}
+}
+
+static void random_test_align_and_overflow(odp_random_kind_t kind)
+{
+	uint8_t ODP_ALIGNED_CACHE buf[64];
+
+	for (int align = 8; align < 16; align++) {
+		for (int len = 0; len < 8; len++) {
```
Comment (psavol) [inline test/validation/api/random/random.c:109]: API does not specify min/max len, but to me it would be an application error if it requests 0 bytes. So, I start validation test from len = 1.

**Review by psavol — APPROVED** [review]
